### PR TITLE
feat(server): Display events involved in the workflow. Closes #3673

### LINKED ISSUE #3673: Surface K8s Events about Nodes to UI
We run argo on GKE with `cluster-autoscaler`. Most of the time all our workflows trigger a `scaleUp` event since powerful machines are usually down when not used.

Usually, the scale-up takes a few minutes. During this time the workflow is in pending state with a message similar to this:

image

Once the node has been created the workflow starts and all is good!

But sometimes the cluster is not able to scale up (often due to an error in the `resources` and `nodeSelector` configuration). 

Admin have access to cluster event logs and we can spot the issue quickly:

image

But most of our staff only use the Argo dashboard and so can only rely on the workflow status message to understand what is going on. And scaling up error reported by the `cluster-autoscaler` is not propagated to it:

```bash
pod didn't trigger scale-up (it wouldn't fit if a new node is added): 16 Insufficient memory, 1 node(s) didn't match node selector, 15 Insufficient cpu
```

Would that be possible to report cluster events related to a specific workflow or pods managed by a workflow in the `status` field?

image

Not sure it's technically possible (because of ServiceAccount and permissions) but I am asking just in case.

### LINKED ISSUE #3725: Show workflows events in the UI like Argo CD does
# Summary

Make it easier to debug workflow failures by surfacing relevant Kubernetes events.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Closes #3673. 
See #2667, #3488

### CODE REVIEW

File: server/workflow/workflow_server.go [hunk 1]
```diff
@@ -201,6 +202,55 @@ func (s *workflowServer) WatchWorkflows(req *workflowpkg.WatchWorkflowsRequest,
 	}
 }
 
+func (s *workflowServer) WatchEvents(req *workflowpkg.WatchEventsRequest, ws workflowpkg.WorkflowService_WatchEventsServer) error {
+	ctx := ws.Context()
```
Comment (alexec) [inline server/workflow/workflow_server.go:206]: copy and paste from `WatchWorkflows`

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 2]
```diff
@@ -136,14 +140,20 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
...
-                                                <WorkflowTimeline
-                                                    workflow={this.state.workflow}
-                                                    selectedNodeId={this.selectedNodeId}
-                                                    nodeClicked={node => this.selectNode(node.id)}
-                                                    ref={timeline => (this.timelineComponent = timeline)}
-                                                />
-                                            )}
+                                            )) ||
+                                                (this.selectedTabKey && (
+                                                    <EventsPanel
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:145]: image

File: ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx [hunk 3]
```diff
@@ -281,6 +282,11 @@ export const WorkflowNodeInfo = (props: Props) => (
                         </div>
                     )
                 },
+                {
+                    title: 'EVENTS',
+                    key: 'events',
+                    content: <EventsPanel namespace={props.workflow.metadata.namespace} fieldSelector={'involvedObject.kind=Pod,involvedObject.name=' + props.node.name} />
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx:288]: image

we should probably include `kind in (Pod,PodDisruptionBudget,PersistentVolumeClaim`

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 2 above)
Comment (alexec) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:145]: we should probably have

`kind in (Workflow,PodDisruptionBudget,PersistentVolumeClaim)`

Comment (alexec) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:145]:
```
[acollins8@intuitdepe1a95 argo (⎈ |k3s-default:argo) (feat-3725)]$ k get event --field-selector='involvedObject.kind in ("Pod")'
Error from server (BadRequest): Unable to find "/v1, Resource=events" that match label selector "", field selector "involvedObject.kind in (\"Pod\")": invalid selector: 'involvedObject.kind in ("Pod")'; can't understand 'involvedObject.kind in ("Pod")'

```

File: ui/src/app/shared/components/resource-editor/resource-editor.tsx [hunk 4]
```diff
@@ -132,9 +133,9 @@ export class ResourceEditor<T> extends React.Component<Props<T>, State> {
             <div>
                 {(this.state.editing && (
                     <>
-                        <label className={`argo-button argo-button--base-o ${this.state.lang}`} style={{marginRight: '5px'}} onClick={e => this.changeLang()} key='data-type'>
-                            {this.state.lang === 'yaml' ? <i className='fa fa-check' /> : <i className='fa fa-times' />} YAML
-                        </label>
+                        <ToggleButton toggled={this.state.lang === 'yaml'} onToggle={() => this.changeLang()}>
```
Comment (alexec) [inline ui/src/app/shared/components/resource-editor/resource-editor.tsx:136]: image

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 5]
```diff
@@ -136,14 +140,17 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
...
-                                                <WorkflowTimeline
-                                                    workflow={this.state.workflow}
-                                                    selectedNodeId={this.selectedNodeId}
-                                                    nodeClicked={node => this.selectNode(node.id)}
-                                                    ref={timeline => (this.timelineComponent = timeline)}
-                                                />
-                                            )}
+                                            )) ||
+                                                (this.selectedTabKey && (
+                                                    <EventsPanel namespace={this.state.workflow.metadata.namespace} kind='Workflow' name={this.state.workflow.metadata.name} />
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:145]: image

image

File: ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx [hunk 6]
```diff
@@ -281,6 +282,11 @@ export const WorkflowNodeInfo = (props: Props) => (
                         </div>
                     )
                 },
+                props.node.type === 'Pod' && {
+                    title: 'EVENTS',
+                    key: 'events',
+                    content: <EventsPanel namespace={props.workflow.metadata.namespace} kind='Pod' name={props.node.name} />
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx:288]: image

**Comment by hadim** [comment]
Does that also fix https://github.com/argoproj/argo/issues/3673 ? Looks like it does to me. Can you confirm?

**Comment by alexec** [comment]
> Does that also fix #3673 ? Looks like it does to me. Can you confirm?

I believe so.

**Comment by alexec** [comment]
build failed on `Install Kustomize`

File: pkg/apiclient/workflow/workflow.proto [hunk 7]
```diff
@@ -146,6 +151,10 @@ service WorkflowService {
         option (google.api.http).get = "/api/v1/workflow-events/{namespace}";
     }
 
+    rpc WatchEvents (WatchEventsRequest) returns (stream k8s.io.api.core.v1.Event) {
+        option (google.api.http).get = "/api/v1/events/{namespace}";
```
Comment (alexec) [inline pkg/apiclient/workflow/workflow.proto:155]: what does Argo CD do?

File: pkg/apiclient/workflow/workflow.proto [hunk 8]
```diff
@@ -146,6 +151,10 @@ service WorkflowService {
         option (google.api.http).get = "/api/v1/workflow-events/{namespace}";
     }
 
+    rpc WatchEvents (WatchEventsRequest) returns (stream k8s.io.api.core.v1.Event) {
+        option (google.api.http).get = "/api/v1/stream/events/{namespace}";
```
Comment (alexec) [inline pkg/apiclient/workflow/workflow.proto:155]: as per Argo CD - `/api/v1/stream`

**Comment by alexec** [comment]
I don't think this is related?

```
Error Trace:	cli_test.go:1019
            				then.go:124
            				cli_test.go:1018
Error:      	Received unexpected error:
            	exit status 1
Test:       	TestCLIWithServerSuite/TestRetryOmit
cli_test.go:1020: 
Error Trace:	cli_test.go:1020
            				then.go:124
            				cli_test.go:1018
Error:      	"time=\"2020-08-18T20:34:51Z\" level=fatal msg=\"rpc error: code = Unknown desc = workflow must be Failed/Error to retry\"\n" does not contain "Status:              Running"
```

**Comment by alexec** [comment]
@jessesuen I've update the HTTP path to match Argo CD as per our discussion. This is ready for review.

**Comment by alexec** [comment]
```
2020-08-25T20:32:15.8981931Z panic: test timed out after 15m0s
```

**Comment by alexec** [comment]
Re-running all jobs.

**Comment by alexec** [comment]
@sarabala1979 this has been sat pending review for 10 days now. I'd like to close out on it (so I can discover what bugs it has!).

We can:

* Merge assuming "lazy concensus"
* Ask @simster7 to review.
* Push to v2.12

**Pull request merged**